internal/pkg/cli/pipeline_deploy.go

```diff
@@ -258,13 +270,70 @@ func (o *deployPipelineOpts) Execute() error {
 		PermissionsBoundary: o.app.PermissionsBoundary,
 	}
 
+	if o.showDiff {
+		tpl, err := o.pipelineStackConfig(deployPipelineInput).Template()
+		if err != nil {
+			return fmt.Errorf("generate the new template for deploy diff: %w", err)
+		}
+		if err = o.diff(deployPipelineInput, tpl); err != nil {
+			var errHasDiff *errHasDiff
+			if !errors.As(err, &errHasDiff) {
+				return err
+			}
+		}
+		if !o.skipConfirmation {
+			contd, err := o.prompt.Confirm(continueDeploymentPrompt, "")
+			if err != nil {
+				return fmt.Errorf("ask whether to continue with the deployment: %w", err)
+			}
+			if !contd {
+				return nil
+			}
+		}
+	}
 	if err := o.deployPipeline(deployPipelineInput); err != nil {
 		return err
 	}
-
 	return nil
 }
 
+// PipelineDeployDiff returns the stringified diff of the template against the deployed template of the pipeline.
+func (o *deployPipelineOpts) pipelineDeployDiff(in *deploy.CreatePipelineInput, template string) (string, error) {
+	tmpl, err := o.pipelineDeployer.Template(stack.NameForPipeline(in.AppName, in.Name, in.IsLegacy))
+	if err != nil {
+		var errNotFound *awscloudformation.ErrStackNotFound
+		if !errors.As(err, &errNotFound) {
+			return "", fmt.Errorf("retrieve the deployed template for %q: %w", in.Name, err)
+		}
+		tmpl = ""
+	}
+	diffTree, err := templatediff.From(tmpl).ParseWithCFNOverriders([]byte(template))
+	if err != nil {
+		return "", fmt.Errorf("parse the diff against the deployed pipeline stack %q: %w", in.Name, err)
+	}
+	buf := strings.Builder{}
+	if err := diffTree.Write(&buf); err != nil {
+		return "", err
+	}
+	return buf.String(), nil
+}
+
+func (o *deployPipelineOpts) diff(in *deploy.CreatePipelineInput, tpl string) error {
```

Comment:
i wonder if we could implement the `templateDiffer` interface (`DeployDiff()`) with `o` and avoid writing our own `diff()` function?